pagespeed/system/controller_manager.cc

```diff
@@ -0,0 +1,228 @@
+  if (pid != 0) {
+    exit(EXIT_SUCCESS);
+  }
+
+  // If we keep the current directory we might keep them from being able to
+  // unmount their filesystem.
+  if (chdir("/") < 0) {
+    handler_->Message(kWarning, "Daemonize: Failed to chdir(/).");
+  }
+
+  // If we disconnect file descriptors then logging will break, so don't.
+}
+
+void ControllerManager::ForkOffControllerProcess() {
+  handler_->Message(kInfo, "Forking controller process off of %d", getpid());
+
+  // Whenever we fork off a controller we save the fd for a pipe to it.  Then if
+  // we fork off another controller we can write a byte to the pipe to tell the
+  // old controller to clean up and exit.
+  static int controller_write_fd = -1;
+  if (controller_write_fd != -1) {
+    // We already forked off a controller earlier.  Tell it to quit by writing a
+    // byte.  If there's no one still with the pipe open we'll get SIGPIPE and
+    // die horribly, but the babysitter-restarting-controller structure makes
+    // that very unlikely.
+    handler_->Message(
+        kInfo, "Writing a byte to a pipe to tell the old controller to exit.");
+    ssize_t status;
+    do {
+      status = write(controller_write_fd, "Q", 1);
+    } while (status == -1 && (errno == EAGAIN ||
+                              errno == EINTR));
+    if (status == -1) {
+      handler_->Message(kWarning, "killing old controller failed: %s",
+                        strerror(errno));
+    }
+  }
+
+  int file_descriptors[2];
+  int pipe_status = pipe(file_descriptors);
+  CHECK(pipe_status != -1) << "Couldn't create a root-controller pipe.";
+
+  pid_t pid = fork();
+  CHECK(pid != -1) << "Couldn't fork a controller babysitter process";
+
+  if (pid != 0) {
+    // Parent process.
+
+    // Close the reading end of the pipe.  We'll never write to it, but when we
+    // (and all our children) die there will be no more processes that could
+    // potentially write to it, and so the people who do have it open for
+    // reading can see that death.
+    close(file_descriptors[0]);
+
+    // Save the writing end of the pipe.
+    controller_write_fd = file_descriptors[1];
+
+    return;
+  }
+
+  // Now we're in the child process.  Set this up as a babysitter process,
+  // that forks off a controller and restarts it if it dies.
+
+  Daemonize();
+  SetUpSignalHandlers();
+
+  factory_->set_is_root_process(false);
+  factory_->PrepareForkedProcess("babysitter");
+
+  // Close the writing end of the pipe.  If we read a byte from the pipe it
+  // means we should quit because a new controller is starting up.  If we get
+  // EOF from the pipe it means we should quit because the master process shut
+  // down.
+  close(file_descriptors[1]);
+  int controller_read_fd = file_descriptors[0];
+
+  handler_->Message(kInfo, "Babysitter running with PID %d", getpid());
+
+  while (true) {
+    pid = fork();
+    CHECK(pid != -1) << "Couldn't fork a controller process";
```

Comment:
Looking over the error codes for fork it really looks to me like if it's failing we're not going to be able to do anything useful:

```
       EAGAIN fork() cannot allocate sufficient memory to  copy  the  parent's
              page tables and allocate a task structure for the child.

           -> We already assume we can always allocate memory, and die if we can't.

       EAGAIN It was not possible to create a new process because the caller's
              RLIMIT_NPROC resource limit was  encountered.   To  exceed  this
              limit,  the  process  must  have either the CAP_SYS_ADMIN or the
              CAP_SYS_RESOURCE capability.

           -> We should not be hitting this limit by creating one new process, but I guess
               we could handle this case.  I'm not sure it's worth it, though.

       ENOMEM fork()  failed  to  allocate  the  necessary  kernel  structures
              because memory is tight.

           -> Same as EAGAIN          

       ENOSYS fork()  is not supported on this platform (for example, hardware
              without a Memory-Management Unit).

           -> Shouldn't apply.
```